src/main/java/io/vertx/core/eventbus/impl/MessageProducerImpl.java

```diff
@@ -127,10 +134,19 @@ public void end() {
   }
 
   @Override
+  public synchronized void end(Handler<AsyncResult<Void>> handler) {
+    endHandler(v -> handler.handle(Future.succeededFuture()));
+    end();
+  }
+
+  @Override
   public void close() {
     if (creditConsumer != null) {
```

Comment:
the code of close should be moved to `end(Handler<AsyncResult<Void>)` instead, so we should have something like

```
public synchronized void end(Handler<AsyncResult<Void>> completionHandler) {
   if (creditConsumer != null) {
      creditConsumer.unregister();
   }
   if (endHandler != null) {
      endHandler.handler(Future.succeededFuture());
   }
}
```

Reply:
note: we should somehow take in account the pending queue before doing the effective end... (another issue)

Reply:
I made `MessageProducerImpl` wait till pending queue is empty before invoking endHandler(). Such implementation kinda invalidates your previous comment: logic is split between `close()` and `handleClosed()`